### PR TITLE
Example Project

### PR DESCRIPTION
**Pull request by WieFel**
- Example project using null-safety

### CODE REVIEW

**Comment by WieFel** [comment]
screen1 screen2

File: example/pubspec.yaml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+# In iOS, build-name is used as CFBundleShortVersionString while build-number used as CFBundleVersion.
+# Read more about iOS versioning at
+# https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html
+version: 1.0.0+1
+
+environment:
+  sdk: ">=2.12.0 <3.0.0"
+
+dependencies:
+  flutter:
```
Comment (calvintam236) [inline example/pubspec.yaml:24]: This line should move under environment `flutter: ">=2.0.0 <3.0.0"`

Comment (WieFel) [reply example/pubspec.yaml:25]: You mean like it is now, right?

**Comment by calvintam236** [comment]
@WieFel can you pull from `master` and update the import here (`package:tinycolor2/tinycolor2.dart`)?

**Review by calvintam236 — CHANGES_REQUESTED** [review]
1. `flutter format` needed
2. merge from `master`

**Comment by WieFel** [comment]
Ok, apparently something went wrong on my rebase before.
I'll just rebase again and force push...

File: example/android/build.gradle [hunk 2]
```diff
@@ -0,0 +1,29 @@
+buildscript {
+    ext.kotlin_version = '1.3.50'
+    repositories {
+        google()
+        jcenter()
```
Comment (calvintam236) [inline example/android/build.gradle:5]: it is deprecated. https://stackoverflow.com/a/67482465

Comment (WieFel) [reply example/android/build.gradle:5]: Ok strange. I just freshly created the project using flutter 2.2.2.
Will replace it

**Review by calvintam236 — CHANGES_REQUESTED** [review]
remove `pubspec.lock` from `example` directory. it is only needed in the repo root for package

**Comment by calvintam236** [comment]
> Ok, apparently something went wrong on my rebase before.
> I'll just rebase again and force push...

btw this is feature branch. it will just squash into one.

**Comment by WieFel** [comment]
>     1. remove `pubspec.lock` from `example` directory. it is only needed in the repo root for package
After removing it, it is regenerated when running the example app. So I think it has to be there... It could be added to `.gitignore`, but I don't see a reason why it should not be commited...

>     2. `flutter format` for the whole `example` directory
Done. No changes at all. Everything was formatted fine.

**Comment by WieFel** [comment]
> btw this is feature branch. it will just squash into one.
Yeah I know. But rebasing keeps the branch network cleaner in my opinion. Also if it is only for the moment...

**Comment by WieFel** [comment]
What we still have to discuss is this: https://github.com/TinyCommunity/tinycolor2/discussions/2#discussioncomment-959454
Let's open a separate discussion topic about example/demo project!?

File: example/lib/widgets/apply_method_item.dart [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    _value = widget.defaultValue;
+    super.initState();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Row(
+      children: [
+        Expanded(
+            child: Padding(
```
Comment (calvintam236) [inline example/lib/widgets/apply_method_item.dart:35]: not sure if your `flutter format` is working. broken like here

**Comment by calvintam236** [comment]
> > ```
> > 1. remove `pubspec.lock` from `example` directory. it is only needed in the repo root for package
> > ```
> 
> 
> After removing it, it is regenerated when running the example app. So I think it has to be there... It could be added to `.gitignore`, but I don't see a reason why it should not be commited...
> 

https://dart.dev/guides/libraries/private-files

File: example/android/build.gradle [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    dependencies {
+        classpath 'com.android.tools.build:gradle:4.1.0'
+        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
+    }
+}
+
+allprojects {
+    repositories {
+        google()
+        jcenter()
```
Comment (calvintam236) [inline example/android/build.gradle:17]: same issue here.

File: example/lib/widgets/apply_method_item.dart [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                InputDecoration(hintText: widget.defaultValue.toString()),
+            onChanged: (value) =>
+                _value = num.tryParse(value) ?? widget.defaultValue,
+          ),
+        )),
+        Padding(
+          padding: const EdgeInsets.symmetric(horizontal: 8.0),
+          child: ElevatedButton(
+              child: Text("Apply ${widget.name}"),
+              onPressed: () => widget.onButtonPressed(_value)),
```
Comment (calvintam236) [inline example/lib/widgets/apply_method_item.dart:48]: also use trailing comma

File: example/lib/widgets/apply_method_item.dart (hunk 3 above)
Comment (WieFel) [reply example/lib/widgets/apply_method_item.dart:35]: It does work. So you mean I should add trailing commas where they are missing!?

File: example/lib/widgets/apply_method_item.dart (hunk 5 above)
Comment (calvintam236) [reply example/lib/widgets/apply_method_item.dart:48]: missing a trailing comma at places like here makes your code difficult to read and review

File: example/lib/widgets/apply_method_item.dart (hunk 3 above)
Comment (calvintam236) [reply example/lib/widgets/apply_method_item.dart]: right now it is inconsistent across the file (some have 2 and some have 4 in the param passing).

look at the next line and you can tell this is off. are you sure everything is spaces? (maybe tab used here?)

**Comment by WieFel** [comment]
> https://dart.dev/guides/libraries/private-files

https://dart.dev/guides/libraries/private-files#pubspeclock
For the example project it is fine to have `pubspec.lock`. We should rather ignore/remove the one  in the project root.

**Review by calvintam236 — DISMISSED** [review]

**Comment by WieFel** [comment]
Example app now looks like this:
screenshot

Should we add also some usages of the extension?
e.g. `Colors.red.lighten(50);`

**Comment by calvintam236** [comment]
> @calvintam236 Example app now looks like this:
> screenshot
> 
> Should we add also some usages of the extension?
> e.g. `Colors.red.lighten(50);`

Yes, that works.

Is it possible to click on the list item to show the code?

**Comment by WieFel** [comment]
> Is it possible to click on the list item to show the code?

You mean show e.g. `Colors.red.lighten(20);` when clicking on "lighten" list item? (e.g. by expanding item)

**Comment by calvintam236** [comment]
> > Is it possible to click on the list item to show the code?
> 
> You mean show e.g. `Colors.red.lighten(20);` when clicking on "lighten" list item? (e.g. by expanding item)

Something like that, yes. It need some hint how to use it without viewing the source code. You can also do it as a dialog. Simpler way is to change `lighten` text to `Color.lighten(20)`.

**Comment by WieFel** [comment]
@calvintam236 ok like this?

screenshot

**Comment by calvintam236** [comment]
> @calvintam236 ok like this? Wanted to keep it as simple as possible.
> 
> screenshot

Shouldn't the rest using `c.` since you have `var c`? (BTW don't use `var`. Specific the type)

It would be great if you can add other constructors such as `.fromRGB` and `.fromString`

**Comment by WieFel** [comment]
> Shouldn't the rest using `c.` since you have `var c`? (BTW don't use `var`. Specific the type)

What do you mean? All the list items DO use `c`.
And ok, I can specify type...

> It would be great if you can add other constructors such as `.fromRGB` and `.fromString`

Ok

**Comment by calvintam236** [comment]
> > Shouldn't the rest using `c.` since you have `var c`? (BTW don't use `var`. Specific the type)
> 
> What do you mean? All the list items DO use `c`.
> And ok, I can specify type...
> 
> > It would be great if you can add other constructors such as `.fromRGB` and `.fromString`
> 
> Ok

I mean in the screenshot. You are using `Colors.blue.xxx` in all of the example.

**Comment by WieFel** [comment]
Yeah, that is an example for the extension usage in a static way.
Of course it could also be `c.color.lighten(20)` for example. Don't know which way is better!?

**Comment by calvintam236** [comment]
> Yeah, that is an example for the extension usage in a static way.
> Of course it could also be `c.color.lighten(20)` for example, but that's almost the same as `c.lighten(20).color`. Don't know which way is better!?

Personally, I use `TinyColor()` over extension because it will not create another instance of `Tinycolor`. I think the extension way is convenience but has performance penalty. It is basically a stripped-down version.

**Comment by WieFel** [comment]
So, should I change `Colors.blue.xxx` to `c.color.xxx`?
In my opinion it should stay `Colors.blue.xxx`.

**Comment by calvintam236** [comment]
> So, should I change `Colors.blue.xxx` to `c.color.xxx`?
> In my opinion it should stay `Colors.blue.xxx`.

you can add another example for original color `Color c = Colors.blue;` and use `c.` for all of the below. it is clear either works

File: example/lib/main.dart [hunk 6]
```diff
@@ -32,64 +32,65 @@ class ExamplePage extends StatelessWidget {
         children: <Widget>[
           _createListItem(
             title: "Original color",
-            subtitle: "var c = TinyColor(Colors.blue)",
-            color: c.color,
+            subtitle:
+                "TinyColor tc = TinyColor(Colors.blue)\nColor c = Colors.blue",
```
Comment (calvintam236) [inline example/lib/main.dart:36]: Instead of `\n`, use `'''`.

File: example/lib/main.dart [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+class TinyColorApp extends StatelessWidget {
+  // This widget is the root of your application.
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      title: 'TinyColor2 Example',
+      theme: ThemeData(
+        primarySwatch: Colors.blue,
```
Comment (calvintam236) [inline example/lib/main.dart:15]: Maybe use grey (matching theme of the logo)?

File: example/lib/main.dart [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      ),
+      home: ExamplePage(),
+    );
+  }
+}
+
+class ExamplePage extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    TinyColor tc = TinyColor(Colors.blue);
```
Comment (calvintam236) [inline example/lib/main.dart:25]: This should be `final`.

File: example/lib/main.dart [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+    return Scaffold(
+      appBar: AppBar(
+        title: Text("TinyColor2 Example"),
```
Comment (calvintam236) [inline example/lib/main.dart:29]: Add `const` for widgets.

File: example/README.md [hunk 10]
```diff
@@ -0,0 +1,16 @@
+# example
+
+A new Flutter project.
+
+## Getting Started
+
+This project is a starting point for a Flutter application.
```
Comment (calvintam236) [inline example/README.md:7]: Update readne. Write instructions how to run this example?

File: example/README.md [hunk 11]
```diff
@@ -1,16 +1,15 @@
...
 
-A new Flutter project.
+This example project shows some basic usages of the `tinycolor2` package. 
 
-## Getting Started
+## Run project
 
-This project is a starting point for a Flutter application.
+### In your favourite  IDE
+Open `tinycolor2/example` in your favourite IDE and run it from there. While `tinycolor2`
```
Comment (calvintam236) [inline example/README.md:8]: I think null safety has merged #24.

Comment (WieFel) [reply example/README.md]: @calvintam236 ok, that went completely unnoticed by me

**Review by calvintam236 — APPROVED** [review]

**Pull request merged**